local Axes = require("./Axes");
export type Axes = Axes.Axes;
local BrickColor = require("./BrickColor");
export type BrickColor = BrickColor.BrickColor;
local CFrame = require("./CFrame");
export type CFrame = CFrame.CFrame;
local Color3 = require("./Color3");
export type Color3 = Color3.Color3;
local ColorSequence = require("./ColorSequence");
export type ColorSequence = ColorSequence.ColorSequence;
local ColorSequenceKeypoint = require("./ColorSequenceKeypoint");
export type ColorSequenceKeypoint = ColorSequenceKeypoint.ColorSequenceKeypoint;
local DateTime = require("./DateTime");
export type DateTime = typeof(DateTime.now());
local Enum = require("./Enum");
export type Enum = Enum.Enum;
local EnumItem = require("./EnumItem");
export type EnumItem = EnumItem.EnumItem;
local Enums = require("./Enums");
export type Enums = Enums.Enums;
local Faces = require("./Faces");
export type Faces = Faces.Faces;
local Font = require("./Font");
export type Font = Font.Font;
local NumberRange = require("./NumberRange");
export type NumberRange = NumberRange.NumberRange;
local NumberSequence = require("./NumberSequence");
export type NumberSequence = NumberSequence.NumberSequence;
local NumberSequenceKeypoint = require("./NumberSequenceKeypoint");
export type NumberSequenceKeypoint = NumberSequenceKeypoint.NumberSequenceKeypoint;
local PhysicalProperties = require("./PhysicalProperties");
export type PhysicalProperties = PhysicalProperties.PhysicalProperties;
local Ray = require("./Ray");
export type Ray = Ray.Ray;
local Rect = require("./Rect");
export type Rect = Rect.Rect;
local UDim = require("./UDim");
export type UDim = UDim.UDim;
local UDim2 = require("./UDim2");
export type UDim2 = UDim2.UDim2;
local Vector2 = require("./Vector2");
export type Vector2 = Vector2.Vector2;
local Vector2int16 = require("./Vector2int16");
export type Vector2int16 = Vector2int16.Vector2int16;
local Vector3 = require("./Vector3");
export type Vector3 = Vector3.Vector3;
local Vector3int16 = require("./Vector3int16");
export type Vector3int16 = Vector3int16.Vector3int16;

return {
    Axes = Axes,
    BrickColor = BrickColor,
    CFrame = CFrame,
    Color3 = Color3,
    ColorSequence = ColorSequence,
    ColorSequenceKeypoint = ColorSequenceKeypoint,
    DateTime = DateTime,
    Enum = Enum,
    EnumItem = EnumItem,
    Enums = Enums,
    Faces = Faces,
    Font = Font,
    NumberRange = NumberRange,
    NumberSequence = NumberSequence,
    NumberSequenceKeypoint = NumberSequenceKeypoint,
    PhysicalProperties = PhysicalProperties,
    Ray = Ray,
    Rect = Rect,
    UDim = UDim,
    UDim2 = UDim2,
    Vector2 = Vector2,
    Vector2int16 = Vector2int16,
    Vector3 = Vector3,
    Vector3int16 = Vector3int16,
};
